### PR TITLE
Add relays and precharge components

### LINKED ISSUE #79: Add support for new component category `Relay`
### What's needed?

Certain microgrids incorporate external relays to enable the connection or disconnection of components within the microgrid. These relays are controlled by writing a binary value (true/false, 1/0) to a digital output of a micro-controller situated within the local microgrid. Subsequently, the digital output operates the external relay, facilitating the switching mechanism. As part of our implementation, we aim to support and facilitate the control of these switching mechanisms to enhance the microgrid's functionality and flexibility.

Please look here for the corresponding issue in the common proto specs: https://github.com/frequenz-floss/frequenz-api-common/issues/43

### Proposed solution

The Microgrid API should facilitate streaming data from Relays. Therefore, it needs data definitions. For an example of such data definitions, please refer to the for inverters [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/inverter.proto).

The data definition should support the following states: `OPENED` and `CLOSED`.

Additionally, the Microgrid API should also support the [`start`](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/e48d52fb96536fc6bb9422c4b2ca666bcc1a6761/proto/frequenz/api/microgrid/microgrid.proto#L233) and [`stop`](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/e48d52fb96536fc6bb9422c4b2ca666bcc1a6761/proto/frequenz/api/microgrid/microgrid.proto#L301) methods for Relays.

Invoking the `start` method should cause a state transition from `OPENED` to `CLOSED`, and the `stop` command should do the reverse.

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### LINKED ISSUE #80: Add support for new component category `Precharge`
### What's needed?

In microgrids, we implement a crucial safety measure called "Precharging" when connecting batteries to inverters. Precharging involves gradually ramping up the DC voltage to prevent any potential damage to sensitive electrical components like capacitors. To achieve this, a precharge circuit operates in parallel to the DC bus.

The process begins by activating the precharge circuit, allowing the voltage to rise gradually. Once the voltage stabilizes at an acceptable level, the main connection is closed, and the precharge circuit is subsequently opened.

While many inverters and batteries come equipped with in-built precharging mechanisms, some may lack this feature. In such cases, we need to ensure full support for external precharge modules within our APIs. This support enables seamless integration and control of external precharge modules, safeguarding electrical components and contributing to the overall reliability and efficiency of the microgrid system.

Please look here for the corresponding issue in the common proto specs: https://github.com/frequenz-floss/frequenz-api-common/issues/44

### Proposed solution

The Microgrid API should facilitate streaming data from Precharge components. Therefore, it needs data definitions. For an example of such data definitions, please refer to the for inverters [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/inverter.proto).

The data definition should support the following states: `PRECHARGING`, `PRECHARGED`, and `INACTIVE`.

Additionally, the Microgrid API should also support the [start](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/e48d52fb96536fc6bb9422c4b2ca666bcc1a6761/proto/frequenz/api/microgrid/microgrid.proto#L233) and [stop](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/e48d52fb96536fc6bb9422c4b2ca666bcc1a6761/proto/frequenz/api/microgrid/microgrid.proto#L301) methods for Precharge components.

Invoking the start method should cause a state transition from `INACTIVE` to `PRECHARGING` to `PRECHARGED`. The stop command should cause a state transition from `PRECHARGED` to `INACTIVE`.

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tiyash-basu-frequenz**
This PR add support for two new components to the Microgrid API:

* Relay
* Precharge Module

### CODE REVIEW

**Review by llucax — COMMENTED** [review]
LGTM but it looks like #103 should be merged first, right? As the release notes mention `StartComponent` / `StopComponent` but the code still uses `Start` and `Stop`.

File: proto/frequenz/api/microgrid/precharge_module.proto
Comment (TalweSingh) [inline proto/frequenz/api/microgrid/precharge_module.proto]: Same here, though you doing it twice makes me wonder if it's on purpose?

**Comment by tiyash-basu-frequenz** [comment]
> LGTM but it looks like https://github.com/frequenz-floss/frequenz-api-microgrid/pull/103 should be merged first, right?

Right, that's the plan. That's why I did not enable auto-merge in these PRs, because I was going to pull these manually.

**Comment by tiyash-basu-frequenz** [comment]
Marked blocked until #103 is merged.

File: proto/frequenz/api/microgrid/relay.proto
Comment (TalweSingh) [inline proto/frequenz/api/microgrid/relay.proto]: Would add a dot after details to be consistent

**Review by TalweSingh — COMMENTED** [review]
LGTM, will approve now since it's just a nitpick.

File: proto/frequenz/api/microgrid/relay.proto
Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/microgrid/relay.proto]: I will add a `.` there. Agree about consistency.

Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/microgrid/relay.proto]: Done

File: proto/frequenz/api/microgrid/precharge_module.proto
Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/microgrid/precharge_module.proto]: Added a `.`.

**Comment by tiyash-basu-frequenz** [comment]
#103 is merged. Unblocked this PR.

File: proto/frequenz/api/microgrid/microgrid.proto
Comment (niklas-timpe) [inline proto/frequenz/api/microgrid/microgrid.proto:228]: is the order o actions important here?

Comment (niklas-timpe) [reply proto/frequenz/api/microgrid/microgrid.proto:228]: Also just for my understanding. Where is a specific relay linked to a specific battery-inverter instance?

Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/microgrid/microgrid.proto:228]: >is the order o actions important here?

No, the documentation is not in any specific order. It just specifies the behaviour for different component categories.

Comment (tiyash-basu-frequenz) [reply proto/frequenz/api/microgrid/microgrid.proto:228]: >Also just for my understanding. Where is a specific relay linked to a specific battery-inverter instance?

It would be a predecessor to components, e.g., inverters. The docs in the relay module should specify this already.

**Review by niklas-timpe — APPROVED** [review]

**Review by niklas-timpe — APPROVED** [review]

File: RELEASE_NOTES.md
Comment (niklas-timpe) [inline RELEASE_NOTES.md:23]: Relays are not just for disconnecting inverters? We could also use relays for other stuff.Is it on purpose to only focus on inverters here?

**Review by TalweSingh — APPROVED** [review]

**Pull request merged**
